unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains_.cpp

```diff
@@ -0,0 +1,506 @@
+/*******************************************************************\
+
+ Module: Unit tests for `instantiate_not_contains_`.
+
+ Author: Jesse Sigal, jesse.sigal@diffblue.com
+
+\*******************************************************************/
+
+#include <catch.hpp>
+
+#include <solvers/refinement/string_constraint_instantiation.h>
+
+#include <solvers/sat/satcheck.h>
+#include <solvers/refinement/bv_refinement.h>
+#include <java_bytecode/java_types.h>
+#include <langapi/mode.h>
+#include <java_bytecode/java_bytecode_language.h>
+#include <util/namespace.h>
+#include <util/symbol_table.h>
+#include <util/simplify_expr.h>
+
+/// \class Types used throughout the test. Currently it is impossible to
+/// statically initialize this value, there is a PR to allow this
+/// diffblue/cbmc/pull/1213
+class tt
+{
+public:
+  tt() {}
+  typet char_type() const {return java_char_type();}
+  typet length_type() const {return java_int_type();}
+  array_typet array_type() const
+  {
+    return array_typet(char_type(), infinity_exprt(length_type()));
+  }
+  refined_string_typet string_type() const
+  {
+    return refined_string_typet(length_type(), char_type());
+  }
+  array_typet witness_type() const
+  {
+    return array_typet(length_type(), infinity_exprt(length_type()));
+  }
+};
+
+// Static variable to get proper types
+const tt t;
+
+/// Creates a `constant_exprt` of the proper length type.
+/// \param [in] i: integer to convert
+/// \return corresponding `constant_exprt`
+constant_exprt from_integer(const mp_integer i)
+{
+  return from_integer(i, t.length_type());
+}
+
+/// Creates a `string_exprt` of the proper string type.
+/// \param [in] str: string to convert
+/// \return corresponding `string_exprt`
+string_exprt make_string_exprt(const std::string &str)
+{
+  const constant_exprt length=from_integer(str.length(), t.length_type());
+  array_exprt content(t.array_type());
+
+  for(const char c : str)
+    content.copy_to_operands(from_integer(c, t.char_type()));
+
+  return string_exprt(length, content, t.string_type());
+}
+
+/// For a constant `string_exprt`, creates a full index set.
+/// \param [in] s: `string_exprt` to create index set for
+/// \return the corresponding index set
+std::set<exprt> full_index_set(const string_exprt &s)
+{
+  mp_integer n;
+  PRECONDITION(!to_integer(s.length(), n));
```

Comment:
We shouldn't use side effect in these macros I think (they could be deactivated by some compilation option in the future, maybe?).